tools/testbench/utils.c

```diff
@@ -256,3 +257,181 @@ void tb_getcycles(uint64_t *cycles)
 	*cycles = 0;
 #endif
 }
+
+static void tb_trim_line(char *new, char *line)
+{
+	int i = 0;
+	int j = 0;
+	int n = strlen(line);
+	bool in_quotes = false;
+
+	/* Trim begin */
+	while (isspace((int)line[i]) && i < n)
+		i++;
+
+	/* Copy line with multiple blanks removed, change single quotes to
+	 * double quotes, don't remove blanks from inside quotes.
+	 */
+	while (i < n) {
+		if (!isspace((int)line[i])) {
+			if (line[i] == '\'')
+				new[j] = '\"';
+			else
+				new[j] = line[i];
+			if (new[j] == '\"')
+				in_quotes = !in_quotes;
+			j++;
+			i++;
+		} else if (in_quotes || ((i + 1) < n && !isspace((int)line[i + 1]))) {
+			new[j] = ' ';
+			j++;
+			i++;
+		} else {
+			i++;
+		}
+	}
+	new[j] = 0;
+}
+
+static int tb_parse_sleep(char *line, int64_t *sleep_ns)
+{
+	char *token = strtok(line, "sleep ");
+
+	*sleep_ns = (int64_t)(atof(token) * 1e9);
+	printf("Info: Next control will be applied after %lld ns.\n", (long long)*sleep_ns);
+	return 0;
+}
+
+int tb_decode_enum(struct snd_soc_tplg_enum_control *enum_ctl, char *token)
+{
+	int enum_items = enum_ctl->items;
+	int i;
+
+	for (i = 0; i < enum_items; i++) {
+		if (strcmp(enum_ctl->texts[i], token) == 0)
+			return i;
+	}
+
+	return -EINVAL;
+}
+
+static struct tb_ctl *tb_find_control_by_name(struct testbench_prm *tp, char *name)
+{
+	struct tb_glb_state *glb = &tp->glb_ctx;
+	struct tb_ctl *ctl;
+	int i;
+
+	for (i = 0; i <  glb->num_ctls; i++) {
+		ctl = &glb->ctl[i];
+		if (strcmp(ctl->name, name) == 0)
+			return ctl;
+	}
+
+	return NULL;
+}
+
+static int tb_parse_amixer(struct testbench_prm *tp, char *line)
+{
+	char *line_end;
+	char *name_str;
+	char *end_str;
+	char control_name[TB_MAX_CTL_NAME_CHARS] = {0};
+	char control_params[TB_MAX_CTL_NAME_CHARS] = {0};
+	char *find_cset_name_str = "cset name=\"";
+	char *find_end_str = "\" ";
+	int find_len = strlen(find_cset_name_str);
+	int find_end_len = strlen(find_end_str);
+	int len;
+	struct tb_ctl *ctl;
+	int ret;
+
+	name_str = strstr(line, find_cset_name_str);
+	if (!name_str) {
+		fprintf(stderr, "error: no control name in script line: %s\n", line);
+		return -EINVAL;
+	}
+
+	end_str = strstr(&name_str[find_len], find_end_str);
+	if (!end_str) {
+		fprintf(stderr, "error: no control name end quote in script line: %s\n", line);
+		return -EINVAL;
+	}
+
+	len = end_str - name_str - find_len;
+	memcpy(control_name, name_str + find_len, len);
+
+	line_end = line + strlen(line);
+	len = line_end - end_str - find_end_len; // Todo rememove linefeed from end
```

Comment:
todo still valid (rememove?)

Reply:
The line feeds were handled, just remove comment.